### PR TITLE
Make bias in lazy modules lazy and avoid create empty tensors

### PR DESCRIPTION
**Pull request by zhuzilin**
Some minor improvement for lazy modules introduced in #44538, #47350 and #51548.

This PR mainly turn the bias to `UninitializedParameter` and instead of creating empty tensors like 
```python
self.bias = Parameter(torch.Tensor(0))
self.bias = UninitializedParameter()
```
I think it would be better to 
```python
self.register_parameter('bias', None)
self.bias = UninitializedParameter()
```

In addition, I change the constructor of the `LazyBatchNorm` from
```python
self.running_mean = UninitializedBuffer()
```
to
```python
self.register_buffer('running_mean', UninitializedBuffer())
```
as the original one would not change the underlying `self._buffers`.

Thank you for your time on reviewing this PR :).

Gently ping @albanD, @mruberry

### CODE REVIEW

File: torch/nn/modules/batchnorm.py [hunk 1]
```diff
@@ -145,13 +145,15 @@ class _LazyBatchNorm(LazyModuleMixin, _BatchNorm):
 
     def __init__(self, eps=1e-5, momentum=0.1, affine=True, track_running_stats=True):
         super(_LazyBatchNorm, self).__init__(
-            0, eps, momentum, affine, track_running_stats)
+            0, eps, momentum, False, False)
```
Comment (albanD) [inline torch/nn/modules/batchnorm.py:148]: I think the two "False" here are worth a comment explaining why we want them.

File: torch/nn/modules/conv.py [hunk 2]
```diff
@@ -1085,16 +1089,19 @@ def __init__(
...
             0,
-            out_channels,
+            0,
             kernel_size,
             stride,
             padding,
             dilation,
             groups,
-            bias,
+            False,
```
Comment (albanD) [inline torch/nn/modules/conv.py:1098]: Same here, a comment about why this is hardcoded to False would be nice!

**Review by albanD — COMMENTED** [review]
Looks ok to me.
It makes the lazy version even more closely linked to the implementation details of the main implementation but I guess that's ok here.

File: torch/nn/modules/batchnorm.py [hunk 3]
```diff
@@ -145,13 +145,15 @@ class _LazyBatchNorm(LazyModuleMixin, _BatchNorm):
...
+        self.affine = affine
+        self.track_running_stats = track_running_stats
         if self.affine:
             self.weight = UninitializedParameter()
             self.bias = UninitializedParameter()
         if self.track_running_stats:
-            self.running_mean = UninitializedBuffer()
-            self.running_var = UninitializedBuffer()
+            self.register_buffer('running_mean', UninitializedBuffer())
```
[the remark this answers is not in the document]
Comment (zhuzilin) [reply torch/nn/modules/batchnorm.py:155]: @albanD This will probably be solved after #35735 is done. And if there isn't anyone who is working on it, I'd love to give it a try~

[the remark this answers is not in the document]
Comment (zhuzilin) [reply torch/nn/modules/batchnorm.py]: Well... I just checked the logic of `register_buffer` and `__setattr__` again and I'm really sorry that I was wrong with this change. The `register_buffer` can only be called once and the subsequent assignment will change the corresponding value. I have changed it back to where is was.

**Comment by zhuzilin** [comment]
@albanD Would you mind taking another look? Thank you~

File: torch/nn/modules/batchnorm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (albanD) [reply torch/nn/modules/batchnorm.py]: > Also, I found that the buffers in _NormBase was set to parameter when track_running_stats is disabled.

Wow, most likely not no!
A quick blame shows that this has been here since this was first introduced so possibly a typo at that time.
cc @jbschlosser can you think of any good reason for us to do that beyond a typo?

File: torch/nn/modules/linear.py [hunk 4]
```diff
@@ -224,5 +229,6 @@ def initialize_parameters(self, input) -> None:  # type: ignore
             with torch.no_grad():
                 self.in_features = input.shape[-1]
                 self.weight.materialize((self.out_features, self.in_features))  # type: ignore
+                self.bias.materialize((self.out_features,))  # type: ignore
```
Comment (albanD) [inline torch/nn/modules/linear.py:232]: Shouldn't there be a check similar to conv here for the case where the users gave bias=False?

**Review by albanD — COMMENTED** [review]
The code looks good to me.

It seems that the doc needs to be updated and we want some basic tests to make sure this works fine no.
That should be good to go after that!

File: torch/nn/modules/linear.py (hunk 4 above)
Comment (zhuzilin) [reply torch/nn/modules/linear.py]: Thank you for the catch! I've updated the  code.

File: torch/nn/modules/batchnorm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbschlosser) [reply torch/nn/modules/batchnorm.py]: Given that `running_mean` / `running_var` are basically the poster children for the use of buffers, I can't think of any reason for them to specifically be parameters! Perhaps `None` wasn't supported for `register_buffer` back then?

**Comment by zhuzilin** [comment]
@albanD I'm not sure where should I update the doc... Could you give me some guidance on that? And as for the test, this pr should not have introduced any new features to the lazy modules and it should work well if the old tests are passed.

**Comment by albanD** [comment]
You can update the doc directly in the docstring of the Lazy versions of the Modules. In particular, you want to mention there that the bias are also lazy now and how we pick the size that they will use.

For the test, you can add that to the existing tests for lazy modules in test/test_nn.py. Just to check that the bias are properly un-initialized before the first forward and then properly initialized after one forward and that they have the right size.

**Comment by zhuzilin** [comment]
@albanD I have updated the doc and added the relevant tests. Could you have another look? :)

File: torch/nn/modules/batchnorm.py [hunk 5]
```diff
@@ -245,9 +250,13 @@ def _check_input_dim(self, input):
...
 
 class LazyBatchNorm1d(_LazyBatchNorm):
-    r"""A :class:`torch.nn.BatchNorm1d` module with lazy initialization of
-    the ``num_features`` argument of the :class:`BatchNorm1d` that is inferred
-    from the ``input.size(1)``.
+    r"""A :class:`torch.nn.BatchNorm1d` module with lazy initialization.
+
+    In this module, the `weight` and `bias` are of :class:`torch.nn.UninitializedParameter`
+    class. And the `running_mean` and `running_var` are of :class:`torch.nn.UnintializedBuffer`.
+    They will be initialized after the first call to ``forward`` is done and the
```
Comment (albanD) [inline torch/nn/modules/batchnorm.py:257]: There is no need to explain that it change after the forward here. This is already explained in the common lazy doc.
Similarly no need to mention the exact types, you can just say that `weight`, `bias`, `running_mean` and `running_var` will be lazily initialized.
And then say explicitly that the num_features will be inferred from the input size.

Maybe something like:
A :class:`torch.nn.BatchNorm1d` module with lazy initialization of
the ``num_features`` argument of the :class:`BatchNorm1d` that is inferred
from the ``input.size(1)``.
The attributes that will be lazily initialized are `weight`, `bias`, `running_mean` and `running_var`.

File: torch/nn/modules/batchnorm.py [hunk 6]
```diff
@@ -347,9 +356,13 @@ def _check_input_dim(self, input):
 
 
 class LazyBatchNorm2d(_LazyBatchNorm):
-    r"""A :class:`torch.nn.BatchNorm2d` module with lazy initialization of
-    the ``num_features`` argument of the :class:`BatchNorm2d` that is inferred
-    from the ``input.size(1)``.
+    r"""A :class:`torch.nn.BatchNorm2d` module with lazy initialization.
+
+    In this module, the `weight` and `bias` are of :class:`torch.nn.UninitializedParameter`
```
Comment (albanD) [inline torch/nn/modules/batchnorm.py:361]: Same comment as above, I thing the old version is more better as it is not repeating what is already in the general doc and to the point.

**Review by albanD — COMMENTED** [review]
Thanks for the update.
The tests look good.
Small comment on the doc.

File: torch/nn/modules/batchnorm.py (hunk 5 above)
Comment (zhuzilin) [reply torch/nn/modules/batchnorm.py]: updated.

**Comment by zhuzilin** [comment]
@albanD I've update the doc.

**Review by albanD — APPROVED** [review]
LGTM!
Thanks for doing all the updates!
